File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-0.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    int goodNodes(TreeNode* root) {
        int res = 0;
        goodNodesHelper(root, -101, res);
        return res;
    }

    void goodNodesHelper(TreeNode *node, int currMax, int &res) {
        if (node == nullptr) return;

        if (node->val >= currMax) {
            res++;
            currMax = node->val;
        }

        goodNodesHelper(node->left, currMax, res);
        goodNodesHelper(node->right, currMax, res);

    }
};

File: Data Structures & Algorithms/depth-of-binary-tree/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        return self.maxDeptRecur(root, 0)

    def maxDeptRecur(self, root, curr_depth):
        if root is None:
            return curr_depth
        
        return max(self.maxDeptRecur(root.left, curr_depth + 1), self.maxDeptRecur(root.right, curr_depth + 1))

        